File: tiqn_app/tiqn_app/report/daily_quantity/daily_quantity.py
# Copyright (c) 2025, IT Team and contributors
# For license information, please see license.txt

import frappe
from frappe import _

def execute(filters=None):
    # Định nghĩa cấu trúc cột
    columns = get_columns()
    
    # Lấy dữ liệu theo bộ lọc
    data = get_data(filters)
    
    return columns, data

def get_columns():
    # Định nghĩa các cột hiển thị trong báo cáo
    return [
        {
            "label": _("Register Date"),
            "fieldname": "register_date",
            "fieldtype": "Date",
            "width": 120
        },
        {
            "label": _("Group"),
            "fieldname": "employee_group",
            "fieldtype": "Data",
            "width": 120
        },
        {
            "label": _("Quantity"),
            "fieldname": "quantity",
            "fieldtype": "Int",
            "width": 100
        }
        
    ]

def get_data(filters):
    # Xây dựng điều kiện truy vấn dựa vào bộ lọc
    conditions = get_conditions(filters)
    
    # Thực hiện truy vấn SQL để lấy và nhóm dữ liệu
    data = frappe.db.sql("""
        SELECT 
            employee_group,
            COUNT(name) as quantity,
            register_date
        FROM 
            `tabVegetarian Meal Detail`
        WHERE
            {conditions}
        GROUP BY 
            employee_group, register_date
        ORDER BY 
            register_date DESC, employee_group
    """.format(conditions=conditions), filters, as_dict=1)
    
    return data

def get_conditions(filters):
    conditions = "1=1"
    
    # Thay đổi từ register_date thành from_date và to_date
    if filters.get("from_date") and filters.get("to_date"):
        conditions += " AND register_date BETWEEN %(from_date)s AND %(to_date)s"
    elif filters.get("from_date"):
        conditions += " AND register_date >= %(from_date)s"
    elif filters.get("to_date"):
        conditions += " AND register_date <= %(to_date)s"
    
    if filters.get("employee_group"):
        conditions += " AND employee_group = %(employee_group)s"
    
    return conditions

@frappe.whitelist()
def get_groups():
    """Lấy danh sách distinct các group từ cơ sở dữ liệu"""
    try:
        groups = frappe.db.sql("""
            SELECT DISTINCT employee_group as value
            FROM `tabVegetarian Meal Detail`
            WHERE employee_group != ''
            ORDER BY employee_group
        """, as_dict=0)
        return [g[0] for g in groups] 
        
    except Exception as e:
        frappe.log_error(f"Error in get_groups: {str(e)}")
        return []

File: tiqn_app/tiqn_app/report/daily_quantity/daily_quantity.js
// Copyright (c) 2025, IT Team and contributors
// For license information, please see license.txt

frappe.query_reports["Daily Quantity"] = {
	"filters": [
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": get_monday_of_current_week(),
			"reqd": 1,
			"on_change": function () {
				let from_date = frappe.query_report.get_filter_value('from_date');
				let to_date = frappe.query_report.get_filter_value('to_date');

				if (from_date && to_date && from_date > to_date) {
					frappe.query_report.set_filter_value('to_date', from_date);
				}
				// Kích hoạt chạy lại báo cáo khi thay đổi giá trị
				frappe.query_report.refresh();
			}
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": get_saturday_of_current_week(),
			"reqd": 1,
			"on_change": function () {
				let from_date = frappe.query_report.get_filter_value('from_date');
				let to_date = frappe.query_report.get_filter_value('to_date');

				if (from_date && to_date && to_date < from_date) {
					frappe.query_report.set_filter_value('from_date', to_date);
				}
				// Kích hoạt chạy lại báo cáo khi thay đổi giá trị
				frappe.query_report.refresh();
			}
		},
		{
			"fieldname": "employee_group",
			"label": __("Group"),
			"fieldtype": "Select",
			"options": "\n", // Bắt đầu với một option trống
			"reqd": 0
		}
	],

	onload: function (report) {
		// Tải danh sách group khi báo cáo được tải
		frappe.call({
			method: "tiqn_app.tiqn_app.report.daily_quantity.daily_quantity.get_groups",
			callback: function (r) {
				if (!r.exc && r.message) {
					let options = [""];
					options = options.concat(r.message || []);

					// Cập nhật options cho filter
					let group_filter = frappe.query_report.get_filter('employee_group');
					group_filter.df.options = options.join("\n");
					group_filter.refresh();
				} else if (r.exc) {
					console.error("Error:", r.exc);
					frappe.msgprint(__("Không thể tải danh sách Group. Lỗi: ") + r.exc);
				}
			}
		});
	}
};

// Hàm lấy ngày thứ 2 (Monday) của tuần hiện tại
function get_monday_of_current_week() {
	let today = frappe.datetime.get_today();
	let day_of_week = moment(today).day(); // 0 là Chủ nhật, 1 là Thứ 2, ..., 6 là Thứ 7

	// Nếu today là Chủ nhật (0), lùi 6 ngày để lấy Thứ 2 tuần trước
	// Nếu today là các ngày khác, lùi (day_of_week - 1) ngày để lấy Thứ 2
	let days_to_subtract = day_of_week === 0 ? 6 : day_of_week - 1;

	return frappe.datetime.add_days(today, -days_to_subtract);
}

// Hàm lấy ngày thứ 7 (Saturday) của tuần hiện tại
function get_saturday_of_current_week() {
	let today = frappe.datetime.get_today();
	let day_of_week = moment(today).day(); // 0 là Chủ nhật, 1 là Thứ 2, ..., 6 là Thứ 7

	// Nếu today là Chủ nhật (0), lùi 1 ngày để lấy Thứ 7 tuần trước
	// Nếu today là Thứ 7 (6), không cần thay đổi
	// Nếu today là các ngày khác, thêm (6 - day_of_week) ngày để lấy Thứ 7
	let days_to_add = day_of_week === 0 ? -1 : (day_of_week === 6 ? 0 : 6 - day_of_week);

	return frappe.datetime.add_days(today, days_to_add);
}